test/remote/instance.go

```diff
@@ -121,20 +123,43 @@ func (i *InstanceInfo) CreateOrGetInstance(imageURL, serviceAccount string) erro
 		Email:  serviceAccount,
 		Scopes: []string{"https://www.googleapis.com/auth/cloud-platform"},
 	}
-	inst.ServiceAccounts = []*compute.ServiceAccount{saObj}
+	newInst.ServiceAccounts = []*compute.ServiceAccount{saObj}
 
 	if pubkey, ok := os.LookupEnv("JENKINS_GCE_SSH_PUBLIC_KEY_FILE"); ok {
 		klog.V(4).Infof("JENKINS_GCE_SSH_PUBLIC_KEY_FILE set to %v, adding public key to Instance", pubkey)
 		meta, err := generateMetadataWithPublicKey(pubkey)
 		if err != nil {
 			return err
 		}
-		inst.Metadata = meta
+		newInst.Metadata = meta
 	}
 
-	if _, err := i.computeService.Instances.Get(i.project, i.zone, inst.Name).Do(); err != nil {
-		op, err := i.computeService.Instances.Insert(i.project, i.zone, inst).Do()
-		klog.V(4).Infof("Inserted instance %v in project: %v, zone: %v", inst.Name, i.project, i.zone)
+	// If instance exists but machine-type doesn't match, delete instance
+	curInst, _ := i.computeService.Instances.Get(i.project, i.zone, newInst.Name).Do()
+	if curInst != nil {
+		if !strings.Contains(curInst.MachineType, newInst.MachineType) {
+			klog.V(4).Infof("Instance machine type doesn't match the required one. Delete instance.")
+			if _, err := i.computeService.Instances.Delete(i.project, i.zone, i.name).Do(); err != nil {
+				return err
+			}
+
+			then := time.Now()
+			err := wait.Poll(15*time.Second, 5*time.Minute, func() (bool, error) {
+				klog.V(2).Infof("Waiting for instance to be deleted. %v elapsed", time.Since(then))
+				if instance, _ = i.computeService.Instances.Get(i.project, i.zone, i.name).Do(); instance != nil {
+					return false, nil
+				}
+				return true, nil
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if _, err := i.computeService.Instances.Get(i.project, i.zone, newInst.Name).Do(); err != nil {
```

Comment:
Can you just check of curInst (or instance) is nil here? I don't think you need an extra Get call. This will also make your API response handling more consistent (right now you're using the first and second parameter in the return call to indicate the same semantics)

Reply:
done.